src/compiler/scala/tools/nsc/typechecker/Typers.scala

```diff
@@ -1122,13 +1124,39 @@ trait Typers extends Adaptations with Tags with TypersTracking with PatternTyper
         }
       }
 
+      // if user wrote case companion C for expected function type, use C.apply or (C.apply _).tupled
+      def adaptApplyInsertion(): Tree =
+        if (currentRun.isScala3Cross && !isPastTyper && tree.symbol != null && tree.symbol.isModule && tree.symbol.companion.isCase && isFunctionType(pt))
+          silent(_.typed(atPos(tree.pos)(Select(tree, nme.apply)))) match {
+            case SilentResultValue(applicator) =>
+              val arity = definitions.functionArityFromType(applicator.tpe)
+              functionOrPfOrSamArgTypes(pt) match {
+                case arg :: Nil if definitions.isTupleType(arg) && arg.typeArgs.lengthCompare(arity) == 0 =>
+                  val tupled = typed(atPos(tree.pos)(Select(applicator, nme.tupled)), mode, pt)
+                  if (!tupled.isErroneous) {
+                    val msg = s"The method `apply` is inserted. The auto insertion will be deprecated, please write `(${tree.symbol.name}.apply _).tupled` explicitly."
+                    context.deprecationWarning(tree.pos, tree.symbol, msg, "2.13.13")
+                    tupled
+                  }
+                  else EmptyTree
+                case args if args.lengthCompare(arity) == 0 =>
+                  val msg = s"The method `apply` is inserted. The auto insertion will be deprecated, please write `${tree.symbol.name}.apply` explicitly."
+                  context.deprecationWarning(tree.pos, tree.symbol, msg, "2.13.13")
+                  typed(atPos(tree.pos)(Select(tree, nme.apply)), mode, pt)
```

Comment:
```suggestion
                  applicator
```
?

Reply:
Gee, I thought I'd tried all the combinations of supplying pt. Using applicator works. So the silent doesn't get pt, and I didn't drill into why that gives me the function; the caveat is that `case class E()` is not expanded; supplying pt fixes that but then the tupled case errors right away. The fix then is if arity < 0, retype with pt for E. I'll add the test for E and also for overloaded apply, which should error nicely. I checked that dotty does apply insert for E. (I understand scala 2 resists eta expanding nilary method.)